flowzone.yml

```diff
@@ -2335,6 +2335,81 @@ jobs:
           echo "sha_tag=${sha_tag}" >> $GITHUB_OUTPUT
           echo "version_tag=${version_tag}" >> $GITHUB_OUTPUT
 
+  python_publish:
+    name: Publish to test PyPI
+    runs-on: ${{ fromJSON(inputs.runs_on) }}
+    timeout-minutes: ${{ fromJSON(inputs.jobs_timeout_minutes) }}
+    needs:
+      - is_python
+      - npm_test
+      - custom_test
+      - docker_test
+      - cargo_test
+      - python_test
+    if: |
+      !failure() && !cancelled() &&
+      needs.npm_test.result == 'success' &&
+      needs.is_python.outputs.python_poetry == 'true'
+    defaults:
+      run:
+        working-directory: .
+        shell: bash --noprofile --norc -eo pipefail -x {0}
+    steps:
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@d27e3f3d7c64b4bbf8e4abfb9b63b83e846e0435 # v4
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install Poetry
+        run: |
+          pipx install poetry
+
+      - name: Run poetry install
+        run: |
+          poetry install
+
+      - name: Publish draft release
+        env:
+          PYPI_TOKEN: ${{ secrets.PYPI_TEST_TOKEN }}
+        run: |
+          poetry config repositories.test-pypi https://test.pypi.org/legacy/
+          poetry config pypi-token.test-pypi $PYPI_TOKEN
+          poetry publish --build -r test-pypi
+
+  python_finalize:
+    name: Finalize npm
+    runs-on: ${{ fromJSON(inputs.runs_on) }}
+    timeout-minutes: ${{ fromJSON(inputs.jobs_timeout_minutes) }}
+    needs:
+      - is_pr_merged
+      - is_pushed_tag
+      - is_npm
+    if: |
+      !failure() && !cancelled() &&
+      needs.is_pr_merged.result != needs.is_pushed_tag.result &&
+      needs.is_python.outputs.python_poetry == 'true'
+    defaults:
+      run:
+        working-directory: .
+        shell: bash --noprofile --norc -eo pipefail -x {0}
+    steps:
+
+      - name: Install Poetry
+        run: |
+          pipx install poetry
+
+      - name: Run poetry install
+        run: |
+          poetry install
+
+      - name: Publish release
+        env:
+          PYPI_TOKEN: ${{ secrets.PYPI_TOKEN }}
+        run: |
+          poetry config pypi-token.pypi $PYPI_TOKEN
+          poetry publish --build
```

Comment:
Generally for finalize steps we do not want to build anything new, and instead try to re-tag or re-upload something built in a previous publish or test step.

For example, docker finalize and balena finalize both just re-tag existing draft artifacts.

NPM finalize cannot re-tag so it downloads the pack from the previous publish run and re-uploads that.

Can pypi.org packages be copied from test.pypi.org to finalize? Or is there some other workflow that would avoid building new artifacts? Maybe following the npm publish & finalize example?

Reply:
hey @klutchell this is still WIP and I am still learning the working of PyPI in more details. After a small struggle this morning using test PyPI I managed to release a pre-release into test PyPI which means I can also do a pre-release on the actual PyPI - so we would probably be able to ditch Test PyPI all together and use only the PyPI (non-test) registry, which would allow for what you described.

FWIW the struggle here is because PyPI does not actually use full semver specification but rather PEP440 which is slightly different: https://peps.python.org/pep-0440/

More over, I can only do pre-releases using the commit hash "small version" such as https://test.pypi.org/project/balena-sdk-python/12.6.0b445977/

Reply:
Sounds like you are on the right track, I'll leave you to it!